src/viewer.js

```diff
@@ -594,72 +859,101 @@ class VolumeImageViewer {
           };
-          if (options.includeIccProfile) {
+          if (this[_includeIccProfile]) {
             retrieveOptions['queryParams'] = {
-              iccprofile: 'yes'
+              iccprofile: 'yes',
             }
           }
-          options.client.retrieveInstanceFramesRendered(retrieveOptions).then(
+
+          this[_options].client.retrieveInstanceFramesRendered(retrieveOptions).then(
             (renderedFrame) => {
-              const blob = new Blob([renderedFrame], {type: mediaType});
-              img.src = window.URL.createObjectURL(blob);
+              img.src = this[_renderingEngine].createURLFromRGBImage(renderedFrame);
             }
           );
         } else {
-          // TODO: support "image/jp2" and "image/jls"
-          const mediaType = 'image/jpeg';
+          // allowed mediaTypes: http://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_8.7.4.html  
+          // we use in order: jls, jp2, jpx, jpeg.
+          const jlsMediaType = 'image/jls';
+          const jlsTransferSyntaxUIDlossless = '1.2.840.10008.1.2.4.80';
+          const jlsTransferSyntaxUID = '1.2.840.10008.1.2.4.81';
+          const jp2MediaType = 'image/jp2';
+          const jp2TransferSyntaxUIDlossless = '1.2.840.10008.1.2.4.90';
+          const jp2TransferSyntaxUID = '1.2.840.10008.1.2.4.91';
+          const jpxMediaType = 'image/jpx';
+          const jpxTransferSyntaxUIDlossless = '1.2.840.10008.1.2.4.92';
+          const jpxTransferSyntaxUID = '1.2.840.10008.1.2.4.93';
+          const jpegMediaType = 'image/jpeg';
+          const jpegTransferSyntaxUID = '1.2.840.10008.1.2.4.50';
           const retrieveOptions = {
             studyInstanceUID,
             seriesInstanceUID,
             sopInstanceUID,
             frameNumbers,
             mediaTypes: [
-              { mediaType, transferSyntaxUID: '1.2.840.10008.1.2.4.50' }
+              { mediaType : jlsMediaType, transferSyntaxUID : jlsTransferSyntaxUIDlossless },
+              { mediaType : jlsMediaType, transferSyntaxUID : jlsTransferSyntaxUID },
+              { mediaType : jp2MediaType, transferSyntaxUID : jp2TransferSyntaxUIDlossless },
+              { mediaType : jp2MediaType, transferSyntaxUID : jp2TransferSyntaxUID },
+              { mediaType : jpxMediaType, transferSyntaxUID : jpxTransferSyntaxUIDlossless },
+              { mediaType : jpxMediaType, transferSyntaxUID : jpxTransferSyntaxUID },
+              { mediaType : jpegMediaType, transferSyntaxUID : jpegTransferSyntaxUID }
             ]
           };
-          options.client.retrieveInstanceFrames(retrieveOptions).then(
+          this[_options].client.retrieveInstanceFrames(retrieveOptions).then(
             (rawFrames) => {
-              const blob = new Blob(rawFrames, {type: mediaType});
-              img.src = window.URL.createObjectURL(blob);
+              img.src = this[_renderingEngine].createURLFromRGBImage(rawFrames[0]);
             }
           );
```

Comment:
This will currently fail for uncompressed images, which is an edge case but one that I already ran into.

Unfortunately, dicomweb-client currently doesn’t let us ask for “Accept: multipart/related; type=application/octet-stream, multipart/related; type=image/*”.

Would the following work?

```js
          this[_options].client.retrieveInstanceFrames(retrieveOptions).then(
            (rawFrames) => {
              img.src = this[_renderingEngine].createURLFromRGBImage(rawFrames[0]);
            }
          ).catch(
            () => {
              const retrieveOptions = {
                studyInstanceUID,
                seriesInstanceUID,
                sopInstanceUID,
                frameNumbers,
                mediaTypes: [
                  { mediaType : 'application/octet-stream' },
                ]
              };
              this[_options].client.retrieveInstanceFrames(retrieveOptions).then(
                (rawFrames) => {
                  img.src = this[_renderingEngine].createURLFromRGBImage(rawFrames[0]);
                }
              )
            }
          );

```

Reply:
yes I have implemented it also for monochorme images.